### DISCUSSION TITLE
panic: Illegal instruction 0x400000003002a073 at pc (0x80000044=>0x80000048).(0=>1): mstatus is not accessible in 0

### DISCUSSION THREAD

**Discussion by qqli579** [gem5/gem5#387] [Q&A] [answered]
When I use the latest gem5(stable branch: commit 48a40cf2f5182a82de360b7efa497d82e06b1631) and use the following command to run coremark：
./build/RISCV/gem5.opt ./configs/deprecated/example/se.py --cmd=./benchmark/coremark.bare.riscv --caches
 I get the following error:
src/arch/riscv/faults.cc:210: panic: Illegal instruction 0x400000003002a073 at pc (0x80000044=>0x80000048).(0=>1): mstatus is not accessible in 0.
At first I thought it was because se.py was abandoned, so I used ./configs/example/gem5_library/arm-hello.py to modify the arm-hello.py file to run with riscv, but the same error was still reported. Can anyone help me as to why this is happening?

**Comment by BobbyRBruce** [comment, marked as answer]
Hey @qqli579 : Only non-privilage, userspace instructions can be simulated in SE mode. Accessing the mstatus register is privileged so that's why you're getting that error. If you want to simulate this you'll need to do it in FS mode.

**Reply by zjh070805** [reply]
When I do it in FS mode and use the following command to run benchmark:
build/RISCV/gem5.opt configs/deprecated/example/fs.py  --script=riscv-tests_src/benchmarks/median.riscv --cpu-type=O3CPU --caches --bp-type=BiModeBP
I get the following error:
src/base/loader/image_file_data.cc:110: fatal: fatal condition fd < 0 occurred: Failed to open file None.
This error typically occurs when the file path specified is incorrect.

**Reply by BobbyRBruce** [reply]
You can't just pass a binary to FS mode, you need to create a disk image for your workload and pass a kernel. I'd suggest using the standard library for this. Examples can be found in "configs/example/gem5_library". A tutorial on using the gem5 stdlib can be found here: https://www.gem5.org/documentation/gem5-stdlib/overview.

I would not rely on the fs.py or se.py scripts. They are deprecated and no longer supported by the project.

**Reply by zjh070805** [reply]
Can I write a simple riscv O3cpu script to run dhrystone like hello world，part of the instructions for disassembling dhrystone are as follows:
    80000044:	3002a073          	csrs	mstatus,t0
    80000048:	4285                	li	t0,1
    8000004a:	02fe                	sll	t0,t0,0x1f
    8000004c:	0002d863          	bgez	t0,8000005c <_start+0x5c>
Why are the following errors still reported：
panic: Illegal instruction 0x400000003002a073 at pc (0x80000048=>0x8000004c).(0=>1): mstatus is not accessible in 0

**Reply by zjh070805** [reply]
Does riscv-tests_src/benckmark only run in FS mode in gem5？

**Reply by BobbyRBruce** [reply]
I'm not entirely sure what you're asking, so I'll explain as much I can about the problem i think your facing.

Yes, you can run drystone in a gem5 FS simulation but you can't _just_ simulate the running of the binary, you must also simulate the entire system inclusive of the operating system. This is the big difference between SE mode and FS mode. SE mode works by using the host OS to handle typical OS functionality thus allowing gem5 to skip the cost of simulating an OS for the application to run in. This isn't costless solution. As SE mode does not simulate the OS the impact of the OS on the simulation system will be unknown and not reflected in the simulation outputs. We leave it to the individual gem5 user to decide if this trade-off works for them. Typically it does if simulated workload, or the region interest within it, is not dependent on the OS in any way.

SE mode i limited moreso in a couple noteworthy ways:

1. gem5 does not guarantee perfect syscall translation to the host. This means certain syscalls compiled into your binary may not work.
2. [The problem you run into] gem5 SE mode only processes non-privileged instructions. This is by design and will never change. Privileged instructions are used by low-level processes in the kernel. As, in SE mode, we are using teh host OS kernel, we cannot do this.

FS mode isn't a drop-in replacement for SE mode. Just passing a binary to FS mode and expecting gem5 to execute it directly is the equivalent to trying to run a binary on a computer system with a wiped disk drive. I.e. no kernel, no bootloader, no filesystem, libraries, utilities, anything. It just won't work. Least of all because the system doesn't know how to load the binary into memory and execute it, but also because the binary probably depends on system calls. For example, if your binary prints to stdout, that's a system call. If it accesses files on the system, that's system call too. In fact, I don't think there's a way to do input/output from a binary without using system calls. System calls are something provided by the OS, so you must simulate it. This is done in gem5 by providing, at minimum a kernel binary and a disk image containing the wider OS.

In your case, what you'd typically do is create a disk image (I'd go for an Ubuntu 22.04 image) and add your benchmark applications to it, to be executed after system boot. Then, when your simulation is run it will boot the entire OS and when that's complete will your binary/binaries be executed. Depending on your simulation configuraiton this could take a very long time. There are tricks we can use to speedup booting (CPU Switching, or Checkpointing), but this does require quite a bit of work. Without these tricks I'd expect booting to take several hours. Using only the O3 CPU it could be over a day to just boot. The O3 CPU very slow.

Examples of creating a riscv disk image can be found here:

* https://github.com/gem5/gem5-resources/tree/stable/src/riscv-ubuntu

**Reply by BobbyRBruce** [reply]
> Does riscv-tests_src/benckmark only run in FS mode in gem5？
 
 No, they run in SE mode. You can see the script used in "tests/gem5/asmtest/configs/riscv_asmtest.py".
 
 Given you've jumped around a bit here with different workloads. What is it you want to do? The riscv-tests are not benchmarks, they just test RISC-V ISA instructions. If you're trying to test the performance of a simulated system this isn't what you want. All these tests do is test the RISC-V ISA implemented correctly (at last for the instructions tested... which is far from all-encompassing).

**Reply by zjh070805** [reply]
Sorry, I did not express clear, I want to run different benchmarks in riscv architecture in gem5, the benchmark code is as follows：
https://github.com/riscv-software-src/riscv-tests/tree/master/benchmarks

I just wrote a simple script like gem5/configs/learning_gem5/part1/simple-riscv.py，but I had the following problem：
panic: Illegal instruction 0x400000003002a073 at pc (0x80000048=>0x8000004c).(0=>1): mstatus is not accessible in 0

So, what I want to ask is, are these benchmarks running only in FS mode, and am I doing this incorrectly？

**Reply by rogerchang23424** [reply]
I think you should try use “configs/example/riscv/fs_linux.py” FS script to run RISC-V benchmark, the “ configs/deprecated/example/fs.py ” is deprecated.

**Reply by powerjg** [reply]
Another problem with using code from the upstream `riscv-tests` repo is that they use an esoteric linking script that calls into the proxy kernel. This is not supported in gem5 in *any* mode.

We support many of the riscv-tests benchmarks in gem5-resources. See https://resources.gem5.org/resources/riscv-tests-dhrystone?version=1.0.0 for dhystone, for instance.

**Reply by yfxiao2** [reply]
I met the same problem. Do you solve it?
